### PR TITLE
fix #1211: non-string headers and derive header name

### LINKED ISSUE #1211: Typesafe Client: Non-String headers
In `HeaderBuilder` line 46 there's a hard cast `(String) parameter.getValue()`. This prevents other header types, e.g. `UUID`, `Long`, etc.

### PR DESCRIPTION
**Pull request by t1**
from parameter or method name

### CODE REVIEW

**Comment by t1** [comment]
I'll look into the failures later today

File: client/implementation/src/main/java/io/smallrye/graphql/client/impl/typesafe/json/JsonReader.java [hunk 1]
```diff
@@ -45,10 +45,6 @@ Object read() {
         if (isListOfErrors(value) && !isGraphQlErrorsType())
             throw cantApplyErrors(readGraphQlClientErrors());
         Reader<?> reader = reader(location);
-        if (reader == null) {
-            throw new InvalidResponseException(
-                    "invalid " + type.getTypeName() + " value for " + location.getDescription() + ": " + value);
-        }
```
Comment (t1) [inline client/implementation/src/main/java/io/smallrye/graphql/client/impl/typesafe/json/JsonReader.java:51]: @jmartisk: these changes are unrelated to this ticket but we discussed about this in another ticket.

File: client/tck/src/main/java/tck/graphql/typesafe/HeaderBehavior.java [hunk 2]
```diff
@@ -53,8 +53,11 @@ void shouldFailToCollectHeadersWithMethodAndConstantHeaders() {
         Throwable thrown = catchThrowable(api::greeting);
 
         then(thrown.getMessage().replace("\"", "")) // JDK 11 prints the quotes, JDK 8 doesn't
-                .isEqualTo("Header with 'method' AND 'constant' not allowed: @" + Header.class.getName()
-                        + "(method=M, constant=C, name=H)");
+                .startsWith("Header with 'method' AND 'constant' not allowed: @" + Header.class.getName() + "(")
+                .contains("method=M")
+                .contains("constant=C")
+                .contains("name=H")
+                .endsWith(")");
```
Comment (t1) [inline client/tck/src/main/java/tck/graphql/typesafe/HeaderBehavior.java:60]: The annotation parameter order seems to vary some times

File: client/tck/src/main/java/tck/graphql/typesafe/HeaderBehavior.java [hunk 3]
```diff
@@ -71,8 +74,11 @@ void shouldFailToCollectHeadersWithMethodNorConstantHeaders() {
         Throwable thrown = catchThrowable(api::greeting);
 
         then(thrown.getMessage().replace("\"", "")) // JDK 11 prints the quotes, JDK 8 doesn't
-                .isEqualTo("Header must have either 'method' XOR 'constant': @" + Header.class.getName()
-                        + "(method=, constant=, name=H)");
+                .startsWith("Header must have either 'method' XOR 'constant': @" + Header.class.getName() + "(")
+                .contains("method=")
+                .contains("constant=")
+                .contains("name=H")
+                .endsWith(")");
```
Comment (t1) [inline client/tck/src/main/java/tck/graphql/typesafe/HeaderBehavior.java:81]: The annotation parameter order seems to vary some times

File: client/tck/src/main/java/tck/graphql/typesafe/HeaderBehavior.java [hunk 4]
```diff
@@ -586,4 +592,81 @@ void shouldAddCharsetRequestAndResponseHeaders() {
...
+        fixture.returnsData("'greeting':'dummy-greeting'");
+        NonStringHeadersApi api = fixture.build(NonStringHeadersApi.class);
+
+        api.greeting(null);
+
+        then(fixture.sentHeader("foo")).isNull();
+    }
+
+    @GraphQLClientApi
+    @Header(method = "foo")
```
Comment (jmartisk) [inline client/tck/src/main/java/tck/graphql/typesafe/HeaderBehavior.java:622]: So the header name will default to the method name? Is that really useful? Http headers names in most cases contain dashes and start with an uppercase letter, which is not quite compatible with naming of methods in Java. I don't see how this is useful in the real world.

File: client/implementation-vertx/src/main/java/io/smallrye/graphql/client/vertx/typesafe/HeaderBuilder.java [hunk 5]
```diff
@@ -58,15 +51,38 @@ public MultiMap build() {
...
+            throw new RuntimeException("Missing header name for constant '" + header.constant() + "'");
+        return new HeaderValue(header.constant(), header.name(), null);
+    }
+
+    private HeaderValue resolve(ParameterInfo parameter) {
+        Header header = parameter.getAnnotations(Header.class)[0];
+        return new HeaderValue(parameter.getValue(), header.name(), parameter.getName());
+    }
+
+    private static class HeaderValue {
```
Comment (jmartisk) [inline client/implementation-vertx/src/main/java/io/smallrye/graphql/client/vertx/typesafe/HeaderBuilder.java:72]: I'd rather call this `HeaderDescriptor` or something, as it contains both the name and value?!

Comment (t1) [reply client/implementation-vertx/src/main/java/io/smallrye/graphql/client/vertx/typesafe/HeaderBuilder.java]: fixed

File: client/tck/src/main/java/tck/graphql/typesafe/HeaderBehavior.java (hunk 4 above)
Comment (t1) [reply client/tck/src/main/java/tck/graphql/typesafe/HeaderBehavior.java:624]: We already convert exception names to kebab-case for the `error-code` field on the server side. I did just that here, too. I find it very convenient to write `@Header(method = "contentType")` and it turns out as `Content-Type`. Same thing for parameter names! WDYT?

Comment (jmartisk) [reply client/tck/src/main/java/tck/graphql/typesafe/HeaderBehavior.java:624]: I still find that a little bit odd, but ok, I won't oppose :) 
One last ask: Could you update `docs/typesafe-client-headers.md` with this?

Comment (t1) [reply client/tck/src/main/java/tck/graphql/typesafe/HeaderBehavior.java:624]: Yes, sure will.

Comment (t1) [reply client/tck/src/main/java/tck/graphql/typesafe/HeaderBehavior.java:624]: done

**Review by jmartisk — APPROVED** [review]

**Comment by jmartisk** [comment]
Thanks!

**Pull request merged**